File: kotlin/src/main/kotlin/378.KthSmallestElementinaSortedMatrix.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/


class KthSmallestElementinaSortedMatrix{
  fun kthSmallest(matrix: Array<IntArray>, k: Int): Int {
    val (n, m) = matrix.size to matrix[0].size
    val que = PriorityQueue<Pair<Int, Int>>(compareBy{ matrix[it.first][it.second] })
    que.add(0 to 0)
    for(x in k-1 downTo 1){
      val (i, j) = que.poll()
      if(j+1 < m) que.add(i to j+1)
      if(j == 0 && i+1 < n) que.add(i+1 to j)
    }
    val (i, j) = que.peek()
    return matrix[i][j]
  }

  @Test
  fun test1(){
    assertEquals(14, kthSmallest(arrayOf(
        intArrayOf(1,5,9),
        intArrayOf(10,11,14),
        intArrayOf(12,13,15)
    ), 8))
  }
}

File: python/378_KthSmallestElementinaSortedMatrix.py
from unittest import TestCase
# https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/
import heapq


class KthSmallestElementinaSortedMatrix(TestCase):
    def kthSmallest(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        n, m = len(matrix), len(matrix[0])

        def count_le(v):
            res, j = 0, m-1
            for i in range(n):
                while j >= 0 and matrix[i][j] > v:
                    j -= 1
                res += j+1
            return res
        l, r = matrix[0][0], matrix[n-1][m-1]
        while l <= r:
            md = (l+r) >> 1
            if count_le(md) < k:
                l = md+1
            else:
                r = md-1
        return l

    def test1(self):
        self.assertEqual(13, self.kthSmallest(
            [
                [1, 5, 9],
                [10, 11, 13],
                [12, 13, 15]
            ], 8
        ))


